### ISSUE TITLE
Masked ``_array2string`` doesn't work on some structured arrays

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#13401]
### Description

The dispatch function ``_array2string `` in ``masked.function_helpers`` doesn't work on arrays with strutcured dtypes.

### Expected behavior

Masked arrays with structured dtypes can be printed no prob, like their non-masked counterparts.

### Actual behavior

It errors because of the structured dtype.

### Steps to Reproduce

Here's an example:

```python
from astropy.utils.masked import Masked
from astropy.uncertainty import Distribution

arr = np.array(Distribution(((np.random.beta(2,5, 10000)-(2/7))/2 + 3)))  # just for convenience.
x = Masked(arr, mask=False)

repr(x)
```

While the following works:

```
repr(arr)
```

### System Details

macOS-10.16-x86_64-i386-64bit
Python 3.9.12 (main, Jun  1 2022, 06:36:29) 
[Clang 12.0.0 ]
Numpy 1.22.3
pyerfa 2.0.0.1
astropy 5.2.dev131+ga2da0589eb.d20220607
Scipy 1.8.1
Matplotlib 3.5.2

**Comment by nstarman** [comment]
@mhvk, I encountered this while looking to see how Masked and Distribution combine. They sort of do, but the subclass generation is not very robust.

**Comment by mhvk** [comment]
@nstarman - does `Masked` work with a regular structured array? (It really should!) If this is `Distribution` specific, I think the issue is the rather lame override of `__repr__` in `Distribution`. Really, `Distribution` should have `__array_function__` defined so this could be done correctly.

Bit more generally, for cases like this I think we are forced to make a choice of which one goes on top. Since a `distribution` has a shape that excludes the samples, the mask in naive usage would just be for each set of samples. I think that a mask for each sample is likely more useful, but that may be tricky...

**Comment by nstarman** [comment]
It does not appear to be Distribution specific. The following fails

```python
q = ((np.random.beta(2,5, 100)-(2/7))/2 + 3) * u.kpc
new_dtype = np.dtype({'names': ['samples'],
                      'formats': [(q.dtype, (q.shape[-1],))]})
q = q.view(new_dtype)
Masked(q)
```

I think it's related to ``q.shape == ()``.

**Comment by mhvk** [comment]
Fix in #13404!
